receiver/filebeatreceiver/filebeat_to_logdata.go

```diff
@@ -0,0 +1,118 @@
+// limitations under the License.
+
+package filebeatreceiver // import "github.com/open-telemetry/opentelemetry-collector-contrib/receiver/filebeatreceiver"
+
+import (
+	"errors"
+	"sort"
+	"time"
+
+	"github.com/elastic/go-lumber/lj"
+	"go.opentelemetry.io/collector/pdata/pcommon"
+	"go.opentelemetry.io/collector/pdata/plog"
+	"go.uber.org/zap"
+)
+
+const (
+	cannotConvertValue = "cannot convert field value to attribute"
+)
+
+// FilebeatToLogData transforms Filebeat events into logs
+func FilebeatToLogData(logger *zap.Logger, batch *lj.Batch) (plog.Logs, error) {
+	ld := plog.NewLogs()
+	rl := ld.ResourceLogs().AppendEmpty()
+	sl := rl.ScopeLogs().AppendEmpty()
+	var instrumentLibrary, instrumentLibraryVersion string
+	for _, event := range batch.Events {
+		logRecord := sl.LogRecords().AppendEmpty()
+		attrValue, _ := event.(map[string]interface{})
+		if _, ok := attrValue["@metadata"]; ok {
+			instrumentLibrary = attrValue["@metadata"].(map[string]interface{})["beat"].(string)
+			instrumentLibraryVersion = attrValue["@metadata"].(map[string]interface{})["version"].(string)
+		}
+		layout := "2006-01-02T15:04:05.000Z"
+		recordTimestamp, timeErr := time.Parse(layout, attrValue["@timestamp"].(string))
+		if timeErr != nil {
+			logger.Debug("Unsupported time conversion", zap.Any("value", event))
+			return ld, errors.New("unsupported time conversion")
+		}
+		logRecord.SetTimestamp(pcommon.NewTimestampFromTime(recordTimestamp))
+
+		for k, v := range attrValue {
+			if k != "message" {
+				err := addAttributeValue(logger, v, logRecord.Attributes().PutEmpty(k))
+				if err != nil {
+					return ld, err
+				}
+			} else {
+				logRecord.Body().SetStr(v.(string))
+			}
+		}
+
+	}
+	sl.Scope().SetName(instrumentLibrary)
+	sl.Scope().SetVersion(instrumentLibraryVersion)
+	return ld, nil
+}
+
+func addAttributeValue(logger *zap.Logger, originalValue interface{}, dest pcommon.Value) error {
+
+	switch value := originalValue.(type) {
+	case nil:
+	case string:
+		dest.SetStr(value)
+	case int64:
+		dest.SetInt(value)
+	case float64:
+		dest.SetDouble(value)
+	case bool:
+		dest.SetBool(value)
+	case map[string]interface{}:
+		return addAttributeMap(logger, value, dest)
+	case []interface{}:
+		return addSliceVal(logger, value, dest)
+	default:
+		logger.Debug("Unsupported value conversion", zap.Any("value", originalValue))
+		return errors.New(cannotConvertValue)
+	}
+	return nil
+}
+
+func addSliceVal(logger *zap.Logger, originalValue []interface{}, dest pcommon.Value) error {
+	arr := dest.SetEmptySlice()
+	for _, elt := range originalValue {
+		err := addAttributeValue(logger, elt, arr.AppendEmpty())
+		if err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func addAttributeMap(logger *zap.Logger, originalValue map[string]interface{}, dest pcommon.Value) error {
+	attrMap := dest.SetEmptyMap()
+	keys := make([]string, 0, len(originalValue))
+	for k := range originalValue {
+		keys = append(keys, k)
+	}
+	sort.Strings(keys)
```

Comment:
Sort of missing this, can you pls explain why the sorting may be required here when both source and and destination are maps?  Or is there some ordering required on the destination?

Reply:
We dont require it. Removed.